### DISCUSSION TITLE
BTHome as an API

### DISCUSSION THREAD

**Discussion by agittins** [wasp-os/wasp-os#518] [Ideas]
I'm interested in trying to get my pinetime to send [BTHome](https://bthome.io) advertisements.

I found a number of earlier issues where people have made initial efforts at altering the advertisements sent by wasp-os but the sticking point (at that time) appears to have been support in micropython. #110  #129 #394 #474 #292 #359 #332 #81 

I'm a huge fan of the BTHome protocol, as it gives:
- optional encryption with PSK
- compact encoding of many types of sensor value
- event support for buttons etc
- low-power, as no BLE connection to track/maintain
- instant, zero-effort creation of usable devices and UI in HomeAssistant

I'm also a huge fan of using bluetooth adverts because HA has bluetooth proxies, so I have sprinkled esp32's all over my house, and any adverts get routed to my HA system via multiple redundant paths.

Apps like https://github.com/thiswillbeyourgithub/SleepTk_pinetime_sleep_tracker, the HR app, and `[insert my planned app here]` might all want to be able to "export" data out of the watch, and having an API that lets apps "register" their sensors to the system, and have them sent out via BTHome would be very cool, and would avoid issues of limited "space" for advertisements. Also would be nice to have things like battery level, step counter etc.

I think that having BTHome adverts available as an api would make a lot of people's ideas easier to implement and more hacker-friendly. I know that @thiswillbeyourgithub is probably keen to see this too, and I want to talk them into exporting sleep-tracking data via the same mechanism :-)

I'm interested in:
- Opinions on how this might be done as an API that could be merged into wasp-os (design, patterns etc)
- Opinions on "where things are at" as far as being able to alter the adverts at all - will this still require c-language changes in upython, is that an upstream thing, does wasp-os use a current version, pointers on where I'd start as far as versions, repos etc. (I'm willing to do the groundwork required, just after some pointers on the higher-level stuff and project-specific context that I might be missing).

I've done some BTHome stuff on nRF51822 using the nordic api, and some other stumbling around in c, so happy to "have at it" once I have a vague idea of the broader strokes of what's worth implementing as far as what wasp-os would benefit from and be able to merge in.

My longer term goal is to write a simple app with one or two screens, each with 4 buttons, with a hard-coded icon. User presses button, BTHome packet goes out, ✨ magic ✨ happens in HomeAssistant. 

The iteration after that would be for the button icons to be able to be updated via the uart connection, so (say) a homeassitant integration for wasp-os(!) could connect to the watch (via a bluetooth proxy) and update the buttons to match the functions available for, say, the room the watch is now in (*cough* https://github.com/agittins/bermuda *cough*)... for example ;-)

But first step is to get some control over the adverts being sent by the device, and an idea of how that might be exposed to apps in a waspish way. All ideas very welcome!

**Comment by jonnor** [comment]
Having better support for Bluetooth advertisements would be great!

One thing to be aware of is that the nrf port of MicroPython uses ubluepy, whereas the other ports are on the "bluetooth" module. There is some work on moving to the same for NRF. Tracking issue at https://github.com/micropython/micropython/issues/9851 - looking for new contributors to land the in-progress MR.
Of course this would also require updating the MicroPython version, so it is a rather large endevour.
